### PR TITLE
Refactor crypto subpackage as discussion in #39

### PR DESCRIPTION
**Pull request by c633**
* See https://github.com/coniks-sys/coniks-go/pull/39#discussion_r73082470

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ hijacking secure communications without getting caught.
 ## Golang Library
 The packages in this library implement the various components of the CONIKS system and may be imported individually.
 
+- ``client``: Client-side protocol implementation
```
Comment (arlolra) [inline README.md]: These changes are undesirable, right.

Comment (c633) [inline README.md]: Right. I fixed it.

File: .gitignore [hunk 2]
```diff
@@ -2,6 +2,10 @@
 *.o
 *.a
 *.so
+*.dylib
+
+# Cgo compiled headers
+libconiks.h
```
Comment (arlolra) [inline .gitignore]: These ones too, no?

File: crypto/util_test.go [hunk 3]
```diff
@@ -0,0 +1 @@
+package crypto
```
Comment (arlolra) [inline crypto/util_test.go]: Can probably drop this ... or write some tests :)

File: crypto/vrf/vrf.go [hunk 4]
```diff
@@ -37,18 +37,21 @@ const (
 	ProofSize        = 32 + 32 + intermediateSize
 )
 
+type PrivateKey [SecretKeySize]byte
+type PublicKey []byte
```
Comment (arlolra) [inline crypto/vrf/vrf.go]: `type PublicKey [PublicKeySize]byte`?

File: .gitignore (hunk 2 above)
Comment (c633) [inline .gitignore]: Oops :(

File: crypto/vrf/vrf.go [hunk 5]
```diff
@@ -37,18 +37,21 @@ const (
 	ProofSize        = 32 + 32 + intermediateSize
 )
 
+type PrivateKey [SecretKeySize]byte
```
Comment (arlolra) [inline crypto/vrf/vrf.go]: Should this be `PrivateKeySize`?

File: crypto/util_test.go (hunk 3 above)
Comment (c633) [inline crypto/util_test.go]: Since @Liamsi wrote test for this, I dropped it.

File: merkletree/pad.go [hunk 6]
```diff
@@ -25,13 +26,13 @@ type PAD struct {
 
 // NewPAD creates new PAD consisting of an array of hash chain
 // indexed by the epoch and its maximum length is len
-func NewPAD(policies Policies, key crypto.SigningKey, len uint64) (*PAD, error) {
+func NewPAD(policies Policies, key sign.PrivateKey, len uint64) (*PAD, error) {
```
Comment (arlolra) [inline merkletree/pad.go]: Can we change `key` to `signKey`?

**Comment by c633** [comment]
I fixed all your comments in 90956a24771840c5c8855a5bc028062d9b8050f8.

File: merkletree/proof.go [hunk 7]
```diff
@@ -16,6 +23,71 @@ type ProofNode interface {
 	Commitment() []byte
 }
 
+func computeLeafHash(ap *AuthenticationPath,
```
Comment (arlolra) [inline merkletree/proof.go]: These proof changes are kind separate and, again, deserve a pull of their own.

Let's try to keep the habit of,

1. Making changes as small as possible;
2. Keeping logically separate changes to their own commits (which kind of follow from 1).

Having lots of small commits that we can merge independently is preferable.  It eases the cognitive load on the reviewer.

Comment (c633) [inline merkletree/proof.go]: Will follow this! Thanks!

**Comment by arlolra** [comment]
You can rebase this now.

**Comment by c633** [comment]
> You can rebase this now.

Done in 5418fb23ed9c8c0858abb75c96c36649f46dea61.

**Comment by arlolra** [comment]
LGTM.  I'll give others a chance to weigh in.

At some point in the future (not necessarily here), it might be worth abstracting this further into interfaces which the `vrf/` and `sign/` packages implement, so that using the `PAD` with other schemes is straightforward.

**Comment by c633** [comment]
> At some point in the future (not necessarily here), it might be worth abstracting this further into interfaces which the vrf/ and sign/ packages implement, so that using the PAD with other schemes is straightforward.

I opened new issue #50 for this.

**Comment by masomel** [comment]
LGTM, too.

**Comment by Liamsi** [comment]
> LGTM, too.

Same here.
